File: 프로그래머스/LV2/소수 찾기.java
import java.util.HashSet;
class Solution {
    public int solution(String numbers) {
        int answer = 0;
        HashSet<Integer> set = new HashSet<>();
        permutation("", numbers, set);
        while(set.iterator().hasNext()) {
            int a = set.iterator().next();
            if(isPrime(a)) answer++;
            set.remove(a);
        }
        return answer;
    }
    
    private boolean isPrime(int num) {
        if(num==2) return true;
        else if(num==0 || num==1 || num%2==0) return false;
        for(int i = 3; i <= (int)Math.sqrt(num); i+=2) {
            if(num%i==0) return false;
        }
        return true;
    }
    
    private void permutation(String prefix, String str, HashSet<Integer> set) {
        int n = str.length();
        if(!prefix.equals("")) set.add(Integer.valueOf(prefix));
        for(int i=0; i<n; i++) {
            permutation(prefix + str.charAt(i), str.substring(0,i) + str.substring(i+1, n), set);
        }
    }
}

File: 프로그래머스/LV2/소수 찾기.py
from itertools import permutations
def solution(numbers):
    numbers = list(numbers)
    per = set()
    for i in range(len(numbers)):
        for j in permutations(numbers, i + 1):
            per_num = int(''.join(j))
            if per_num not in [0,1]: per.add(per_num)
            for n in range(2, int(per_num ** 0.5) + 1):
                if per_num % n == 0:
                    per.remove(per_num)
                    break
    return len(per)
